### PR TITLE
Implementation of Kwon and Zou Data-OOB: Out-of-bag Estimate as a Sim…

### PR DESCRIPTION
**Pull request by BastienZim**
…ple and Efficient Data Value ICML 2023 using pyDVL

### Description

This PR adds the implementation of a data valuation method described in Kwon and Zou "Data-OOB: Out-of-bag Estimate as a Simple and Efficient Data Value" published at ICML 2023.

The notebook provided gives a comprehensive overview of the method, through examples, visualizations and point removal-evaluation. 

No unit tests were added, as the notebook is testing the method. If-ever that is considered useful, I could write some.

### Changes

- Created compute_data_oob which computed the data_oob data values. The function is compatible with all sklearn estimators as weak learners, provided through the utility model parameter. 

### Checklist

- [ ] Wrote Unit tests (if necessary)
- [x] Updated Documentation (if necessary)
- [ ] Updated Changelog
- [x] If notebooks were added/changed, added boilerplate cells are tagged with `"tags": ["hide"]` or `"tags": ["hide-input"]`

### CODE REVIEW

**Review by mdbenito — COMMENTED** [review]
Salut Bastien and thanks a lot for your first contribution! Data-OOB is an interesting approach and it's great that you chose to help out implementing it. 🙏🏽 

I have left a few comments below about the main module. We can do the notebook in a second step if you want.

One general comment is that you will want to run the linter and type checker. Basically run `tox -e linting` and `tox -e type-checking` to look for errors. The first will rearrange module imports, run black and so on. The second will, well... run the type checker 😄 . You will also want to run mkdocs to see how your documentation renders. See the file [CONTRIBUTING.md](https://github.com/aai-institute/pyDVL/blob/develop/CONTRIBUTING.md) for more info

Finally, if you want to run the removal experiment faster in your notebook, you can use this code. There's a bit of boilerplate, but you can use as many cores as you have by setting `max_workers`.

```python
from tqdm.notebook import tqdm
from concurrent.futures import FIRST_COMPLETED, wait
from pydvl.utils.parallel import init_executor

def removal_job(method: str, n_est=300, max_samples=0.95, progress=False):
    if method == "random":
        values = ValuationResult.from_random(size=len(utility.data))
    else:
        values = compute_data_oob(u=utility, n_est=n_est, max_samples=max_samples, progress=progress)

    best_scores = compute_removal_score(
                u=utility,
                values=values,
                percentages=removal_percentages,
                remove_best=True,
                )
    best_scores["method_name"] = values.algorithm

    worst_scores = compute_removal_score(
            u=utility,
            values=values,
            percentages=removal_percentages,
            remove_best=False,
            )
    worst_scores["method_name"] = values.algorithm
    
    return best_scores, worst_scores

all_best_scores = []
all_worst_scores = []

removal_percentages = np.arange(0, 0.99, 0.01)

n_runs = 20
pending = set()

with init_executor(max_workers=24) as executor:
    for i in range(n_runs):
        pending.add(executor.submit(removal_job, method="random"))
        pending.add(executor.submit(removal_job, method="data_oob"))

    pbar = tqdm(total=2*n_runs, unit="%")
    while len(pending) > 0:
        pbar.n = 2*n_runs - len(pending) + 1  # HACK
        pbar.refresh()
        completed, pending = wait(
                pending, timeout=config.wait_timeout, return_when=FIRST_COMPLETED
                )
        for future in completed:
            best_scores, worst_scores = future.result()
            all_best_scores.append(best_scores)
            all_worst_scores.append(worst_scores)

best_scores_df = pd.DataFrame(all_best_scores)
worst_scores_df = pd.DataFrame(all_worst_scores)
```

File: src/pydvl/value/oob/oob.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+from pydvl.utils.status import Status
+from pydvl.value.result import ValuationResult
+import numpy as np
+from sklearn.ensemble import BaggingClassifier
+
+__all__ = ["compute_data_oob"]
+
+
+def compute_data_oob(
+    u: Utility, n_est: int = 10, max_samples: float = 0.8, *, progress: bool = True
```
[the remark this answers is not in the document]
Comment (BastienZim) [reply src/pydvl/value/oob/oob.py:20]: I will add the support for regression models.

File: src/pydvl/value/oob/oob.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    """
+
+    n_samples = len(u.data)
+    all_counts, all_means, all_M2 = (
+        np.zeros(n_samples),
+        np.zeros(n_samples),
+        np.zeros(n_samples),
+    )
+
+    bag = BaggingClassifier(u.model, max_samples=max_samples, n_estimators=n_est)
```
[the remark this answers is not in the document]
Comment (BastienZim) [reply src/pydvl/value/oob/oob.py:39]: Agreed

File: src/pydvl/value/oob/oob.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    bag.fit(u.data.x_train, u.data.y_train)
+    #    print(bag.score(u.data.x_test, u.data.y_test))
+    for est, samples in maybe_progress(
+        zip(bag.estimators_, bag.estimators_samples_), progress
+    ):  # The bottleneck is the bag fitting not this part so TQDM is not very useful here
+        oob_idx = np.intersect1d(u.data.indices, np.unique(samples))
```
[the remark this answers is not in the document]
Comment (BastienZim) [reply src/pydvl/value/oob/oob.py:45]: It is an obvious mistake on my part, and np.setxor1d should do the job. I also find it quite intriguing that the results do not differ that much. I might be because the base learners have not overfitted too much the data they were trained on, so evaluating them on their train set isn't that bad.

File: src/pydvl/value/oob/oob.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    variances, sample_variances = all_M2 / all_counts, all_M2 / (all_counts - 1)
+
+    return ValuationResult(
+        algorithm="data_oob",
+        status=Status.Converged,
+        values=all_means,
+        data_names=u.data.data_names,
+        counts=all_counts,
+        variances=variances,
+    )
```
[the remark this answers is not in the document]
Comment (BastienZim) [reply src/pydvl/value/oob/oob.py:66]: It greatly simplifies the code. I had not found that method by exploring the library code.

File: src/pydvl/value/oob/oob.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        np.zeros(n_samples),
+        np.zeros(n_samples),
+        np.zeros(n_samples),
+    )
+
+    bag = BaggingClassifier(u.model, max_samples=max_samples, n_estimators=n_est)
+    bag.fit(u.data.x_train, u.data.y_train)
+    #    print(bag.score(u.data.x_test, u.data.y_test))
+    for est, samples in maybe_progress(
+        zip(bag.estimators_, bag.estimators_samples_), progress
```
[the remark this answers is not in the document]
Comment (BastienZim) [reply src/pydvl/value/oob/oob.py:43]: I did not find any mean to add the bagging model fitting time, for large datasets this fitting time can be long. The user could benefit having that time or its ETA displayed.

[the remark this answers is not in the document]
Comment (BastienZim) [reply src/pydvl/value/oob/oob.py:43]: The paper authors indicate that the complexity of data-oob is roughly the fitting of the bagging model.

[the remark this answers is not in the document]
Comment (mdbenito) [reply src/pydvl/value/oob/oob.py]: Maybe two progress bars? The first one is a dummy around the fitting, with no idea when it's going to finish, but at least saying, hey something's happening here. Or maybe a log message.

File: src/pydvl/value/oob/oob.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
+# Code written by @BastienZim
+# Bastien Zimmermann - bastien.zimmermann@craft.ai
```
Comment (AnesBenmerzoug) [inline src/pydvl/value/oob/oob.py:2]: I think it's not good practice to put contributor names directly in the code. That puts additional pressure on other contributors and may prevent them from making to changes to that code. On top of that, there is already a [Github feature](https://github.com/aai-institute/pyDVL/graphs/contributors) that allows people to see the contributors and what contributions they made.

**Comment by BastienZim** [comment]
There still remains a problem raised by the type-checker that I did not manage to resolve.
It concerns the return type of compute_data_oob. 
"Returning Any from function declared to return "ValuationResult[Any, Any]""

I would be interested in knowing the solution.

File: src/pydvl/value/oob/oob.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+            u.model, n_estimators=n_est, max_samples=max_samples, n_jobs=n_jobs
+        )
+        if loss is None:
+            loss = point_wise_accuracy
+    else:
+        bag = BaggingRegressor(
+            u.model, n_estimators=n_est, max_samples=max_samples, n_jobs=n_jobs
+        )
+        if loss is None:
+            loss = neg_l2_distance
```
Comment (mdbenito) [inline src/pydvl/value/oob/oob.py:91]: This is a bit nit-picking (sorry), but you only use `model_type` once and following the same pattern, so you could simply do:
```suggestion

    if is_classifier(u.model):
        bag = BaggingClassifier(
            u.model, n_estimators=n_est, max_samples=max_samples, n_jobs=n_jobs
        )
        if loss is None:
            loss = point_wise_accuracy
    elif is_regressor(u.model):
        bag = BaggingRegressor(
            u.model, n_estimators=n_est, max_samples=max_samples, n_jobs=n_jobs
        )
        if loss is None:
            loss = neg_l2_distance
    else:
        raise Exception(
            "Model has to be a classifier or a regressor in sklearn format."
        )
```

File: src/pydvl/value/oob/oob.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+        preds: Model prediction on
+        y:  data labels corresponding to the model predictions
+
+    Returns:
+        Array of point wise accuracy
+    """
+    return np.array(
+        preds == y,
+        dtype=bool,
+    )
```
Comment (mdbenito) [inline src/pydvl/value/oob/oob.py:124]: Why not return a numeric type? (it's a characteristic function)
```suggestion
    return np.array(preds == y, dtype=np.int_)
```

File: src/pydvl/value/oob/oob.py [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+
+
+def neg_l2_distance(preds: NDArray, y: NDArray) -> NDArray[np.float64]:
```
Comment (mdbenito) [inline src/pydvl/value/oob/oob.py:127]: This function is actually generic. With 

```python
T = TypeVar("T", bound=np.number)
```
you can do

```suggestion
def neg_l2_distance(preds: NDArray[T], y: NDArray[T]) -> NDArray[T]:
```

and then use `preds.dtype` as the return value's `dtype`.

File: docs/assets/pydvl.bib
Comment (mdbenito) [inline docs/assets/pydvl.bib:1]: Thanks for adding this!

**Review by mdbenito — COMMENTED** [review]
I added some superficial comments. Looking good!

File: src/pydvl/value/oob/oob.py [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+        max_samples: The fraction of samples to draw to train each base estimator.
+        n_jobs: The number of jobs to run in parallel used in the bagging procedure for both fit and predict.
+        loss: A function taking as parameters model prediction and corresponding data labels(preds, y) and returning an array of point wise errors.
+        progress: If True, display a progress bar.
+
+    Returns:
+        Object with the data values.
+    """
+
+    result = ValuationResult.empty(
```
[the remark this answers is not in the document]
Comment (BastienZim) [reply src/pydvl/value/oob/oob.py]: Ok, thanks for the explanation.

File: src/pydvl/value/oob/oob.py (hunk 7 above)
Comment (BastienZim) [reply src/pydvl/value/oob/oob.py]: It's definitely a good idea. I originally added the neg_l2_distance in an if(is_classifier)/else part ; that is why I had this two parts structure. I forgot to simplify the code upon adding the loss parameter.

File: src/pydvl/value/oob/oob.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+    T is a score function that represents the goodness of a weak learner $\hat{f}_b$ at the i-th datum $(x_i, y_i)$.
+
+    There is a need to tune n_est and max_samples jointly to ensure all samples are at least 1 time oob, otherwise the result could include a nan value for that datum.
+
+    Args:
+        u: Utility object with model, data, and scoring function.
+        n_est: Number of estimator used in the bagging procedure.
+        max_samples: The fraction of samples to draw to train each base estimator.
+        n_jobs: The number of jobs to run in parallel used in the bagging procedure for both fit and predict.
+        loss: A function taking as parameters model prediction and corresponding data labels(preds, y) and returning an array of point wise errors.
```
[the remark this answers is not in the document]
Comment (BastienZim) [reply src/pydvl/value/oob/oob.py:60]: There seem to be some problems with my VScode formatter, I'll investigate this one.

File: src/pydvl/value/oob/oob.py (hunk 8 above)
Comment (BastienZim) [reply src/pydvl/value/oob/oob.py:121]: I have barely no idea of how type combination are supposed to be chosen. My reasoning was that the bool was the cheapest to store memory-wise. But if numpy is more efficient for the following computations with an int type, then 'np.int_' is better.

File: src/pydvl/value/oob/oob.py [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+
+def neg_l2_distance(preds: NDArray, y: NDArray) -> NDArray[np.float64]:
+    r"""Computes negative l2 distance between label and model prediction
+
+    Args:
+        preds: Model prediction on
+        y:  data labels corresponding to the model predictions
+
+    Returns:
+        Array with point wise negative l2 distance between label and model prediction
```
Comment (AnesBenmerzoug) [inline src/pydvl/value/oob/oob.py:129]: Since you call `np.square` before returning the array then all the values are positive. Why is it called *negative* l2 distance?

Comment (BastienZim) [reply src/pydvl/value/oob/oob.py:132]: A mistake on my part. Thanks Anes. FYI the negative part is there to match the classification values order. The higher the value, the better the result.

File: src/pydvl/value/oob/oob.py [hunk 13]
```diff
@@ -0,0 +1,139 @@
+"""
+## References
+
+[^1]: <a name="kwon_data_2023"></a>Kwon et al. [Data-OOB: Out-of-bag Estimate as a Simple and Efficient Data Value]
+(https://proceedings.mlr.press/v202/kwon23e.html). In: Published at ICML 2023
```
Comment (AnesBenmerzoug) [inline src/pydvl/value/oob/oob.py:5]: For markdown links to be rendered properly there shouldn't a new line in the middle.

```suggestion
[^1]: <a name="kwon_data_2023"></a>Kwon et al.
[Data-OOB: Out-of-bag Estimate as a Simple and Efficient Data Value](https://proceedings.mlr.press/v202/kwon23e.html).
In: Published at ICML 2023
```

File: src/pydvl/value/oob/oob.py [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+
+
+def compute_data_oob(
+    u: Utility,
+    n_est: int = 10,
+    max_samples: float = 0.8,
+    n_jobs: int = None,
+    loss: Callable = None,
+    *,
+    progress: bool = True,
```
Comment (AnesBenmerzoug) [inline src/pydvl/value/oob/oob.py:34]: For almost all methods we set this to False be default to avoid cluttering the terminal:

```suggestion
    progress: bool = False,
```

**Comment by AnesBenmerzoug** [comment]
Hi @BastienZim it looks good. It's almost ready there are just a few things missing:

- In order to render the notebook in the docs, you have to manually add `nav` entry to the [mkdocs.yml](https://github.com/aai-institute/pyDVL/blob/develop/mkdocs.yml) file under `nav -> Data Valuation -> Examples`.
- In the notebook's introduction, you pasted the raw link to the publication. I think you should put that on the name of the paper just like you did in the docstring.
- Most matplotlib plotting functions return objects and they are printed out in jupyter notebooks. You should add a semi-colon `;` at the end of the last line of a plotting call e.g. `ax.set_xlabel("Point rank");` or assign the result to `_` e.g. `_ = plt.plot(np.arange(len(oob_values.values)), oob_values.values)`
- We use tags on jupyter notebook cells to customize documentation rendering. We either hide an entire cell using the tag `"hide"` like import cells and the tag "hide-input" for plotting cells to hide the plotting code. Could you please add that? Refer to the Data Utility Learning notebook for an example.

**Review by AnesBenmerzoug — DISMISSED** [review]
@BastienZim Thanks for your work! It looks good to me now.
@mdbenito I think it's ready to be merged. All that's left to do is to add a changelog entry and maybe a description for the oob module. What do you think?

File: mkdocs.yml [hunk 15]
```diff
@@ -199,6 +199,7 @@ nav:
       - KNN Shapley: examples/shapley_knn_flowers.ipynb
       - Data utility learning: examples/shapley_utility_learning.ipynb
       - Least Core: examples/least_core_basic.ipynb
+      - Data oob: examples/data_oob.ipynb
```
Comment (AnesBenmerzoug) [inline mkdocs.yml:202]:
```suggestion
      - Data OOB: examples/data_oob.ipynb
```

**Review by mdbenito — APPROVED** [review]
Agreed that this is ready. We can polish any cosmetic details later post-merge. Thanks Bastien for the contribution, looking forward to more! :)
